### DISCUSSION TITLE
The live streaming Speech Text web socket is not getting connected

### DISCUSSION THREAD

**Discussion by ManjuJain-aviso** [deepgram/community#770] [General help] [answered]
### Which Deepgram product feature are you using?

Deepgram API - STT Streaming

### Details

The live streaming Speech Text web socket is not getting connected
trying to create a WebSocket proxy in node server 

### If you are making a request to the Deepgram API, what is the full Deepgram URL you are making a request to?

wss://api.deepgram.com/v1/listen?model=nova-2&smart_format=true

### If possible, please attach your code or paste it into the text box.

const wss = new WebSocket.Server({ server, path: 'deepgram' });
    // model=nova-2&smart_format=true&language=en-US&sample_rate=16000&encoding=linear16&interim_results=truevad_events=true&utterance_end_ms=1000&endpointing=300

    wss.on('connection', function connection (ws) {
        const socket = new WebSocket('wss://api.deepgram.com/v1/listen?model=nova-2-general&smart_format=true&language=en-US&sample_rate=16000&encoding=linear16',
            {
                headers: {
                    // Remember to replace the YOUR_DEEPGRAM_API_KEY placeholder with your Deepgram API Key
                    Authorization: `Token ${DEEPGRAM_API_KEY}`
                }
            });

        ws.send(JSON.stringify({ status: 'pending' }));

        // Close the WebSocket connection when done
        function closeWebSocket () {
            const closeMsg = JSON.stringify({ type: 'CloseStream' });
            socket.send(closeMsg);
        }

        let refreshIntervalId = null;
        socket.onopen = () => {
            ws.send(JSON.stringify({ status: 'connected' }));

            // Send KeepAlive messages every 3 seconds
            refreshIntervalId = setInterval(() => {
                const keepAliveMsg = JSON.stringify({ type: 'KeepAlive' });
                socket.send(keepAliveMsg);
                console.log('Sent KeepAlive message');
            }, 3000);
        };

        socket.onmessage = (message) => {
            const received = JSON.parse(message.data);
            console.log('text', received.channel.alternatives);
            ws.send(received.channel.alternatives);
        };

        socket.onerror = (err) => {
            console.log(err);
        };

        ws.on('message', function incoming (message) {
            if (socket.readyState === WebSocket.OPEN) {
                console.log(message);
                socket.send(message);
            } else {
                if (socket.readyState === 3) {
                    socket.close();
                    ws.close();
                    clearInterval(refreshIntervalId);
                }
            }
        });

        ws.on('close', function close () {
            console.warn('Client disconnected');
            clearInterval(refreshIntervalId);
            socket.close();
            closeWebSocket();
        });

        ws.on('error', function error (err) {
            clearInterval(refreshIntervalId);

            if (err.message === 'Could not send. Connection not open.') {
                ws.close();
                closeWebSocket();
            }
        });
    });

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Reply by ManjuJain-aviso** [reply]
request_id: eb4a662c-28cc-47d5-abe9-dd3fbfe8f381
shared the WebSocket URL and code snippets
didn't connect to WebSocket so **request_id** not mentioned  below error that we are getting 
<img width="1353" alt="Screenshot 2024-05-21 at 12 58 00 AM" src="https://github.com/deepgram/community/assets/103415750/677d81bc-519b-4b5e-b427-28c5a8889f05">

**Comment by SandraRodgers** [comment]
Hi @ManjuJain-aviso ,

Here's some code that works for me. Could you try following this example and see how it works for you?

```js
const WebSocket = require("ws");

const authToken = "DEEPGRAM_API_KEY"; // Replace 'DEEPGRAM_API_KEY' with your actual authorization token
const headers = {
  Authorization: `Token ${authToken}`,
};

// Initialize WebSocket connection
const ws = new WebSocket("wss://api.deepgram.com/v1/listen", { headers });

// Handle WebSocket connection open event
ws.on("open", function open() {
  console.log("WebSocket connection established.");

  // Send audio data (replace this with your audio streaming logic)
  // Example: Read audio from a microphone and send it over the WebSocket
  // For demonstration purposes, we're just sending a KeepAlive message

  setInterval(() => {
    const keepAliveMsg = JSON.stringify({ type: "KeepAlive" });
    ws.send(keepAliveMsg);
    console.log("Sent KeepAlive message");
  }, 3000); // Sending KeepAlive messages every 3 seconds
});

// Handle WebSocket message event
ws.on("message", function incoming(data) {
  console.log("Received:", data);
  // Handle received data (transcription results, errors, etc.)
});

// Handle WebSocket close event
ws.on("close", function close() {
  console.log("WebSocket connection closed.");
});

// Handle WebSocket error event
ws.on("error", function error(err) {
  console.error("WebSocket error:", err.message);
});

// Gracefully close the WebSocket connection when done
function closeWebSocket() {
  const closeMsg = JSON.stringify({ type: "CloseStream" });
  ws.send(closeMsg);
}

// Call closeWebSocket function when you're finished streaming audio
// For example, when user stops recording or when the application exits
// closeWebSocket();
```

**Reply by ManjuJain-aviso** [reply]
Hi @SandraRodgers 
I have written the web socket  proxy in the node server to handle speech to text, from UI we are sending linear16ArrayBuffer but  transcripts are empty here is the  request_id: **eb4a662c-28cc-47d5-abe9-dd3fbfe8f381**

` /*
 *WebSocket Handler.
 */

const dotenv = require('dotenv');
const WebSocket = require('ws');

dotenv.config();

const DEEPGRAM_API_KEY = '';
module.exports = function (server) {
    const wss = new WebSocket.Server({ server, path: `/mic` });
   

    wss.on('connection', function connection (ws) {
        const socket = new WebSocket('wss://api.deepgram.com/v1/listen?model=nova-2-ea&smart_format=true&language=en-US&sample_rate=16000&encoding=linear16&interim_results=true',
            {
                headers: {
                    // Remember to replace the YOUR_DEEPGRAM_API_KEY placeholder with your Deepgram API Key
                    Authorization: `token ${DEEPGRAM_API_KEY}`
                }
            });

        ws.send(JSON.stringify({ status: 'pending' }));

        // Gracefully close the WebSocket connection when done
        function closeWebSocket () {
            const closeMsg = JSON.stringify({ type: 'CloseStream' });
            socket.send(closeMsg);
        }

        let refreshIntervalId = null;
        socket.onopen = () => {
            ws.send(JSON.stringify({ status: 'connected' }));
            // Send KeepAlive messages every 3 seconds
            refreshIntervalId = setInterval(() => {
                const keepAliveMsg = JSON.stringify({ type: 'KeepAlive' });
                socket.send(keepAliveMsg);
                console.log('Sent KeepAlive message');
            }, 3000);
        };

        socket.onmessage = (message) => {
            const received = JSON.parse(message.data);

            console.log('text', received.channel.alternatives);

            if (received.is_final) {
                console.log('Final', received.channel.alternatives);
                ws.send(JSON.stringify({ transcript: received.channel.alternatives[0].transcript }));
            }
        };

        socket.onerror = (err) => {
            console.log(err);
        };

        ws.on('message', function incoming (message) {
            if (socket.readyState === WebSocket.OPEN) {
                console.log(message);

                // sending data format from UI linear16ArrayBuffer
                socket.send(message);
            } else {
                console.log('Socket is not ready', socket.readyState);
                if (socket.readyState === 3) {
                    socket.close();
                    ws.close();
                    clearInterval(refreshIntervalId);
                }
            }
        });

        ws.on('close', function close () {
            console.warn('Client disconnected');
            clearInterval(refreshIntervalId);
            socket.close();
            closeWebSocket();
        });

        ws.on('error', function error (err) {
            clearInterval(refreshIntervalId);
            console.error('WebSocket error:', err);

            if (err.message === 'Could not send. Connection not open.') {
                // Handle the error here (e.g., log it, close the connection, etc.)
                ws.close();
                closeWebSocket();
            }
        });
    });
};
`

**Comment by SandraRodgers** [comment]
@ManjuJain-aviso,

It looks like you need to set the binaryType of the WebSocket to 'arraybuffer' in your Node.js WebSocket server to handle the binary audio data correctly. Here's the updated version of your code with ws.binaryType set:

```js
const wss = new WebSocket.Server({ server, path: /mic });

    wss.on('connection', function connection (ws) {
        ws.binaryType = 'arraybuffer'; // Set binaryType to arraybuffer

        const socket = new WebSocket('wss://api.deepgram.com/v1/listen?model=nova-2-ea&smart_format=true&language=en-US&sample_rate=16000&encoding=linear16&interim_results=true',
            {
                headers: {
                    Authorization: `token ${DEEPGRAM_API_KEY}`
                }
            });

        ws.send(JSON.stringify({ status: 'pending' }));

        // Gracefully close the WebSocket connection when done
        function closeWebSocket () {
            const closeMsg = JSON.stringify({ type: 'CloseStream' });
            socket.send(closeMsg);
        }

        let refreshIntervalId = null;
        socket.onopen = () => {
            ws.send(JSON.stringify({ status: 'connected' }));
            // Send KeepAlive messages every 3 seconds
            refreshIntervalId = setInterval(() => {
                const keepAliveMsg = JSON.stringify({ type: 'KeepAlive' });
                socket.send(keepAliveMsg);
                console.log('Sent KeepAlive message');
            }, 3000);
        };

```

By setting ws.binaryType = 'arraybuffer';, the WebSocket server will correctly handle the binary data being sent from the UI. This should resolve the issue of empty transcripts.

**Reply by ManjuJain-aviso** [reply]
Hey @SandraRodgers 
thanks for your support, but still getting this response after adding the binaryType  **request_id**: 'a9f2abe6-e469-48cb-ab0c-3547c666ada9',
`{
  type: 'Metadata',
  transaction_key: 'deprecated',
  request_id: 'a9f2abe6-e469-48cb-ab0c-3547c666ada9',
  sha256: 'be166ba976eb6de351cae5f2b22f320d6d7f30e5a68282e62d3eaed41e9b5013',
  created: '2024-05-24T12:20:36.925Z',
  duration: 0,
  channels: 0
}`

kind of transaction_key: 'deprecated',

**Reply by davidvonthenen** [reply, marked as answer]
hi @ManjuJain-aviso 

It seems like you have an issue following the defined protocol. I would review these:
https://developers.deepgram.com/docs/lower-level-websockets
https://developers.deepgram.com/reference/listen-live

If you aren't keen on using one of the SDKs that performs all this for you, you can use the SDKs as a template for a known working example.
